src/main/java/com/aws/greengrass/clientdevices/auth/api/DomainEvents.java

```diff
@@ -5,19 +5,34 @@
 
 package com.aws.greengrass.clientdevices.auth.api;
 
+import com.aws.greengrass.logging.api.Logger;
+import com.aws.greengrass.logging.impl.LogManager;
+
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.Consumer;
 
 
 @SuppressWarnings({"rawtypes","unchecked"})
 public class DomainEvents {
     private final Map<Class, CopyOnWriteArrayList<DomainEventListener>> eventListeners = new ConcurrentHashMap<>();
+    private final Logger logger = LogManager.getLogger(DomainEvents.class);
+
+    private Consumer<Result<? extends  Exception>> errorHandler;
+
+    public DomainEvents() {
+        this.errorHandler = (Result<? extends Exception> result) -> {};
+    }
 
     @FunctionalInterface
     public interface DomainEventListener<T extends DomainEvent> {
-        void handle(T event);
+        Result<?> handle(T event);
+    }
+
+    public void onError(Consumer<Result<? extends Exception>> consumer) {
+        this.errorHandler = consumer;
```

Comment:
There are probably better ways to signal back to the service that an error has occurred. This requires the failure to have been triggered by an event, which I don't think will always be the case. Why not just introduce a critical failure event that the service can subscribe to?